dist-material/bin/oapService.sh

```diff
@@ -40,10 +40,14 @@ OAP_OPTIONS=" -Doap.logDir=${OAP_LOG_DIR}"
 eval exec "\"$_RUNJAVA\" ${JAVA_OPTS} ${OAP_OPTIONS} -classpath $CLASSPATH org.apache.skywalking.oap.server.starter.OAPServerStartUp \
         2>${OAP_LOG_DIR}/oap.log 1> /dev/null &"
 
+PID="$!"
+OAP_PID_FILE="${OAP_HOME}/bin/oap.pid"
+
 if [ $? -eq 0 ]; then
```

Comment:
`$?` should be also extracted into variable

Reply:
like this
```shell
OAP_PID_FILE="${OAP_HOME}/bin/oap.pid"
eval exec "\"$_RUNJAVA\" ${JAVA_OPTS} ${OAP_OPTIONS} -classpath $CLASSPATH org.apache.skywalking.oap.server.starter.OAPServerStartUp \
        2>${OAP_LOG_DIR}/oap.log 1> /dev/null & /bin/echo -n "$$" > "$OAP_PID_FILE""
```

Reply:
> like this
> 
> ```shell
> OAP_PID_FILE="${OAP_HOME}/bin/oap.pid"
> eval exec "\"$_RUNJAVA\" ${JAVA_OPTS} ${OAP_OPTIONS} -classpath $CLASSPATH org.apache.skywalking.oap.server.starter.OAPServerStartUp \
>         2>${OAP_LOG_DIR}/oap.log 1> /dev/null & /bin/echo -n "$$" > "$OAP_PID_FILE""
> ```

It's ok, but you have to remove the PID file if `$? != 0`, right?

Reply:
I'm sorry , i don't test before.


> > like this
> > ```shell
> > OAP_PID_FILE="${OAP_HOME}/bin/oap.pid"
> > eval exec "\"$_RUNJAVA\" ${JAVA_OPTS} ${OAP_OPTIONS} -classpath $CLASSPATH org.apache.skywalking.oap.server.starter.OAPServerStartUp \
> >         2>${OAP_LOG_DIR}/oap.log 1> /dev/null & /bin/echo -n "$$" > "$OAP_PID_FILE""
> > ```
> 
> It's ok, but you have to remove the PID file if `$? != 0`, right?

I'm sorry , i don't test before, this script can't get correct PID. tomcat shell script 
```shell
  else
    eval $_NOHUP "\"$_RUNJAVA\"" "\"$LOGGING_CONFIG\"" $LOGGING_MANAGER "$JAVA_OPTS" "$CATALINA_OPTS" \
      -D$ENDORSED_PROP="\"$JAVA_ENDORSED_DIRS\"" \
      -classpath "\"$CLASSPATH\"" \
      -Dcatalina.base="\"$CATALINA_BASE\"" \
      -Dcatalina.home="\"$CATALINA_HOME\"" \
      -Djava.io.tmpdir="\"$CATALINA_TMPDIR\"" \
      org.apache.catalina.startup.Bootstrap "$@" start \
      >> "$CATALINA_OUT" 2>&1 "&"

  fi

  if [ ! -z "$CATALINA_PID" ]; then
    echo $! > "$CATALINA_PID"
  fi
```

Reply:
@kezhenxu94 Because $! or $$ is used after exec, the PID obtained is the PID of the shell of the child process executed by exec, so the definition of $! can only be defined after the execution of exec. The reference tomcat's startup.sh is also written like this .

```shell
  if [ ! -z "$CATALINA_PID" ]; then
    echo $! > "$CATALINA_PID"
  fi
```

Reply:
@kezhenxu94 Because $! or $$ is used after exec, the PID obtained is the PID of the shell of the child process executed by exec, so the definition of $! can only be defined after the execution of exec. The reference tomcat's startup.sh is also written like this .

```shell
  if [ ! -z "$CATALINA_PID" ]; then
    echo $! > "$CATALINA_PID"
  fi
```